### PR TITLE
DDIM sampling in training example + storing training data as images

### PR DESCRIPTION
**Pull request by lukovnikov**
In this PR, I slightly changed the DDIM scheduler and pipeline in the main code and added a sampling script in unconditional generation example, as well as a script to save images from datasets as they would be used for training such models.

For the DDIM scheduler, the timestep variable now contains pairs of x_t and x_tm1, rather than having to compute x_tm1 inside the step method (the previous implementation won't work if we want to add support for non-linear (e.g. quadratic) selection of subsets of steps).

Also, support for x0-prediction mode is added in DDIM sampling code.

### CODE REVIEW

File: src/diffusers/pipelines/ddim/pipeline_ddim.py [hunk 1]
```diff
@@ -68,6 +69,9 @@ def __call__(
                 [PIL](https://pillow.readthedocs.io/en/stable/): `PIL.Image.Image` or `np.array`.
             return_dict (`bool`, *optional*, defaults to `True`):
                 Whether or not to return a [`~pipeline_utils.ImagePipelineOutput`] instead of a plain tuple.
+            substeps_mode (`str`, *optional*, defaults to "linear"):
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/ddim/pipeline_ddim.py:72]: Hmm this is a bit too hacky for me and not really related to DDIM anymore IMO. DDIM is defined as selecting substeps linearly IMO. Do we really need this?

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 2]
```diff
@@ -161,6 +163,16 @@ def __init__(
         self.num_inference_steps = None
         self.timesteps = torch.from_numpy(np.arange(0, num_train_timesteps)[::-1].copy().astype(np.int64))
 
+    @classmethod
+    def from_ddpm_scheduler(cls, sched):
+        ret = cls(
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:169]: Let's please not introduce a new class method API

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 3]
```diff
@@ -185,26 +197,40 @@ def _get_variance(self, timestep, prev_timestep):
 
         return variance
 
-    def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.device] = None):
+    def set_timesteps(
+        self,
+        num_inference_steps: int,
+        device: Union[str, torch.device] = None,
+        substeps_mode: str = "linear",
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:205]: As mentioned above, I'm not sure we want to add this to the native DDIM scheduler

**Comment by patrickvonplaten** [comment]
Sorry I don't really understand how the training script relates to the changes in the DDIM scheduler and pipeline

File: src/diffusers/pipelines/ddim/pipeline_ddim.py (hunk 1 above)
Comment (lukovnikov) [reply src/diffusers/pipelines/ddim/pipeline_ddim.py:72]: We probably don't need this, but in the [DDIM paper](https://arxiv.org/pdf/2010.02502.pdf), they explored a non-linear option (see Appendix D.2 on p.18) and in principle you can select the steps however you want.
Should I remove this since it's not supported yet anyway?

**Comment by patrickvonplaten** [comment]
Hey @lukovnikov,

Thanks for the PR, sorry I'm a bit confused what this PR is about - is it about changes to the DDIM scheduler or training scripts? It would be great to try to keep PRs laser focused on one feature addition :-) 

Regarding adding more functionality to what substeps for DDIM should be chosen, I'm not sure it's a good idea to add this to DDIM since DDIM is defined as using linearly downscaled timesteps IMO

**Comment by lukovnikov** [comment]
> Thanks for the PR, sorry I'm a bit confused what this PR is about - is it about changes to the DDIM scheduler or training scripts? It would be great to try to keep PRs laser focused on one feature addition :-)

Thanks for reviewing, I'll change this PR to training example changes only and make another one with the other changes.

> Regarding adding more functionality to what substeps for DDIM should be chosen, I'm not sure it's a good idea to add this to DDIM since DDIM is defined as using linearly downscaled timesteps IMO

In the DDIM paper (in appendix) they also mention a non-linear schedule and the author told me that in principle you can select any steps you want but i'll remove this since it's not supported anyway.
